python/ray/data/aggregate.py

```diff
@@ -33,14 +36,53 @@ def __init__(self,
                 two accumulators into one.
             finalize: This is called once to compute the final aggregation
                 result from the fully merged accumulator.
+            on: The data subset on which to compute the aggregation.
+                For a simple dataset: it can be a callable.
+                For an Arrow dataset: it can be a column name or a callable.
+                The default is to compute the aggregation on the entire row.
             name: The name of the aggregation. This will be used as the output
-                column name in the case of Arrow dataset.
+                column name in the case of Arrow dataset. If not given, then
+                f"{type(self).__name__.lower()}({on})" will be used.
         """
         self.init = init
-        self.accumulate = accumulate
+        self._accumulate = accumulate
         self.merge = merge
         self.finalize = finalize
-        self.name = name
+        self.on = on
+        self._name = name
+
+    @property
+    def name(self):
+        name = self._name
+        if name is None:
+            name = type(self).__name__.lower()
+            if self.on is not None:
+                name += f"({self.on})"
+            else:
+                name += "()"
+        return name
+
+    def accumulate(self, a: AggType, r: T) -> AggType:
+        # Accumulate on data subset.
+        on_fn = _to_on_fn(self.on)
+        return self._accumulate(a, on_fn(r))
+
+    def with_on(self, on: AggregateOnT):
+        """Return a copy of this aggregation with the provided `on` data subset
+        selector.
+
+        Args:
+            on: The data subset on which to compute the aggregation.
+                For a simple dataset: it can be a callable.
+                For an Arrow dataset: it can be a column name or a callable.
+                The default is to compute the aggregation on the entire row.
+
+        Returns:
+            A copy of this aggregation with the provided `on` set.
+        """
+        new_agg = copy.copy(self)
+        new_agg.on = on
+        return new_agg
```

Comment:
Not super happy with this design, but didn't want to introduce another user-facing abstraction, even if the `on`/`name` lifetimes don't quite line up with the rest of `AggregateFn`. Ideally, we'd have an `Aggregation` class that would hold `on` and `name` along with an `AggregateFn` instance (which would just have the `init`, `accumulate`, `merge`, `finalize` fields).